### ISSUE TITLE
In-memory sessions for bots (bypassing Redis)

### ISSUE THREAD

**Issue by fbrnc** [colinmollenhour/Cm_RedisSession#91]
Why do we need to create a real session for bots in the first place? My idea is to detect that we're dealing with a bot in the constructor and then bypass Redis altogether keeping the current "Session" in memory (as simple class property). After all the bot will not be keeping track of the session cookie anyway and the next request (even within the shorter bot session lifetime) would result in a new session.

@colinmollenhour, unless you can already think of a good reason why this is a bad idea I'd like to prepare a PR for you. Does this sound interesting? (It would save us a ton of orphan sessions and reduce communication with Redis significantly - especially with a load balancers in place that will do health checks every couple of seconds...)

**Comment by colinmollenhour** [comment]
This is an almost identical request to #87. I get where you're coming from but I think the risk is pretty high for false positives which the current bot regex. That is, high enough that having no session at all might not be a good idea. My recommendation was to just disable locking using the defined constant but still use the storage backend in case the match is a false positive. You can also set the "bot_lifetime" to be extremely low to further reduce storage but it's still going to be a pretty much constant amount of storage.

Another angle is that if you take the session backend out of the LB checks then if a node can't access the session backend it will never get marked as down by the LB. I'm not sure how ELB works so this might not apply.

Lastly, I think something like this really belongs at a higher level such as Mage_Core_Model_Session_Abstract_Varien. Looking at the start method I see that it returns immediately `$_SESSION` is set so in your bootstrap you could init $_SESSION using your black hole backend and get the desired effect.

Btw, a few easy modifications to Mage_Core_Model_Session_Abstract_Varien can be made to reduce duplication of the validator data to save some session storage overhead.

**Comment by fbrnc** [comment]
Hi @colinmollenhour,
thanks for your super-fast response. I agree this could be risky and that the session backend might not necessarily be the right place to implement this.
I do still think that bypassing Redis in the first place (since even the short lifetimes will add stress to Redis) is a significant improvement. 
As you suggested I'll probably look into hooking this in on a different level (after all this is not specific to the Redis cache backend). Also I'd probably want the bot detection to be a lot less aggressive (or at least configurable separately), (Really what I want is 5 instances with a ELB healthcheck running every 5 seconds not to generate 60 new orphan sessions per minute :)

**Comment by colinmollenhour** [comment]
Yeah, that's a lot of checks but my point was just that if the lifetime is only say 1 minute then you never have more than 60 wasted sessions so storage is negligible.

Just an interesting idea I had but haven't had time to implement in case you are interested since it sounds like you are trying to handle some massive traffic: Sessions without carts are a lot less important as far as them expiring so one way to maximize customer experience (returning to their cart weeks later) while minimizing storage would be to increase the session lifetime only if it is linked to a quote. E.g. normal lifetime might be 7 days and lifetime for sessions with a quote could be bumped to 60 days. Should be pretty easy I think but I haven't thought it through or tried it out yet.

**Comment by fbrnc** [comment]
Very interesting thought!

**Comment by s-hoffman** [comment]
@fbrnc, In line with @colinmollenhour comment regarding `Mage_Core_Model_Session_Abstract_Varien`:
You should be able to overwrite `Mage_Core_Model_Session `to blackhole/local the session.
`Mage_Core_Model_Session_Abstract_Varien::start()` appears to support '**user**' as a session save method (Line 60).

Roughly:
Overwrite `getSessionSaveMethod` to return '**user**' if you find a bot.
Overwrite `getSessionSavePath` to return the name of your custom `session_set_save_handler `function setter.
(See `Mage_Core_Model_Resource_Session::setSaveHandler` for example; blackhole will just /dev/null as needed).
I have not tested this at all though it looks promising, assuming your happy with your bot detection algorithm.

Also, and more Hacky:
`getSessionSaveMethod` and `getSessionSavePath` both reach into the config to get the current value.
(Mage_Core_Model_Session_Abstract::XML_NODE_SESSION_SAVE,
Mage_Core_Model_Session_Abstract::XML_NODE_SESSION_SAVE_PATH
You can create an observer or some other rewrite, to modify the config values in flight.
This method does less damage to the session section of the code at the expense of dependency on the specific method used. (Here, that the system config is used). (Also, a timing dependency with respect to the session and config modify method).

**Comment by fbrnc** [comment]
Hi,

quick follow-up: I created a new module https://github.com/AOEpeople/Aoe_BlackHoleSession prototyping this behavior. It's still experimental, but if this works well we might roll it out on production soon (with a regex only matching ELB and Varnish health checks - since the rest is being handled CloudFront/Varnish anyway).

Thanks @s-hoffman for suggesting an implementation. That's pretty much exactly how I did it (rewriting `Mage_Core_Model_Session`)
